File: DigitalImageProcessing/python/open_close_morphology.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Funciton: 形态学操作：开运算(腐蚀后膨胀)和闭运算(膨胀后腐蚀); 顶帽变换、底帽变换、形态学梯度
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-10-12
"""

import sys

import cv2 as cv


# --------------------------
if __name__ == "__main__":
    if len(sys.argv) > 1:
        image = cv.imread(sys.argv[1], 0)
        if image is None:
            print(f"Error: no such file or dictory.")
        cv.imshow("OriginImage",image)

        # 结构元半径和迭代次数以及最大值
        r, i = 1, 1
        MAX_R, MAX_I = 20, 20

        # 显示形态学处理效果的窗口
        cv.namedWindow("morphology", 1)
        # 回调函数
        def nothing(*arg):
            pass
        # 调节结构元半径
        cv.createTrackbar("r", "morphology", r, MAX_R, nothing)
        # 调节迭代次数
        cv.createTrackbar("i", "morphology", r, MAX_I, nothing)

        while True:
            # 获取当前进度条上的 r 半径值
            r = cv.getTrackbarPos("r", "morphology")
            # 获取当前进度条上的 i 迭代次数
            i = cv.getTrackbarPos("i", "morphology")
            # 创建结构元
            s = cv.getStructuringElement(cv.MORPH_RECT, (2*r+1, 2*r+1))

            # 1. 进行形态学处理——开运算
            # d = cv.morphologyEx(image, cv.MORPH_OPEN, s, iterations=i)

            # 2. 进行形态学处理——闭运算
            # d = cv.morphologyEx(image, cv.MORPH_CLOSE, s, iterations=i)

            # 3. 进行形态学处理——顶帽变换
            # d = cv.morphologyEx(image, cv.MORPH_TOPHAT, s, iterations=i)

            # 4. 进行形态学处理——底帽变换
            # d = cv.morphologyEx(image, cv.MORPH_BLACKHAT, s, iterations=i)

            # 5. 进行形态学处理——形态学梯度
            d = cv.morphologyEx(image, cv.MORPH_GRADIENT, s, iterations=i)

            # 可视化处理效果
            cv.imshow("morphology", d)
            # cv.imwrite("./image/open.png", d)
            ch = cv.waitKey(5)
            # Esc 退出循环
            if ch == 27:
                break

        cv.waitKey(0)
        cv.destroyAllWindows()

    else:
        print("Usge: python.py imageFile")

File: DigitalImageProcessing/cpp/open_close_morphology.cpp
/**
 * @File    : open_close_morphology.cpp
 * @Brief   : 形态学操作：开运算(腐蚀后膨胀)和闭运算(膨胀后腐蚀); 顶帽变换、底帽变换、形态学梯度
 * @Author  : Wei Li
 * @Date    : 2021-10-13
*/

#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

// 开运算和闭运算
int r = 1;                       // 结构元半径
int i = 1;                      // 迭代次数
int Max_R = 20;                // 设置最大半径
int Max_I = 20;               // 设置最大迭代次数
cv::Mat morphologyImg;       // 膨胀后图像
cv::Mat image;              // 输入图像

// 回调函数，调节半径 r 和迭代次数 i
void callBack(int, void*)
{
    // 创建矩形结构元
    cv::Mat s = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(2*r+1, 2*r+1));
    // 进行开运算和闭运算
    // cv.morphologyEx (src, dst, op, kernel, anchor = new cv.Point(-1, -1), iterations = 1, 
                        // borderType = cv.BORDER_CONSTANT, borderValue = cv.morphologyDefaultBorderValue())
    // 1. 开运算
    // cv::morphologyEx(image, morphologyImg, cv::MORPH_OPEN, s, cv::Point(-1, -1), i);

    // 2. 闭运算
    // cv::morphologyEx(image, morphologyImg, cv::MORPH_CLOSE, s, cv::Point(-1, -1), i);

    // 3. 顶帽变换
    // cv::morphologyEx(image, morphologyImg, cv::MORPH_TOPHAT, s, cv::Point(-1, -1), i);

    // 4. 底帽变换
    // cv::morphologyEx(image, morphologyImg, cv::MORPH_BLACKHAT, s, cv::Point(-1, -1), i);

    // 5. 形态学梯度
    cv::morphologyEx(image, morphologyImg, cv::MORPH_GRADIENT, s, cv::Point(-1, -1), i);

    cv::imshow("morphology", morphologyImg);
}

// -------------------------------
int main(int argc, char **argv)
{
    image = cv::imread(argv[1], 0);
    if (!image.data)
    {
        std::cout << "Error: Reading image file unsuccessfully.";
        return -1;
    }
    cv::imshow("OriginImage", image);

    // 创建显示膨胀效果的窗口
    cv::namedWindow("morphology", 1);
    // 创建调节 半径 radio 的进度条
    cv::createTrackbar("r", "morphology", &r, Max_R, callBack);
    cv::createTrackbar("i", "morphology", &i, Max_I, callBack);
    callBack(0, 0);

    cv::waitKey(0);
    cv::destroyAllWindows();
    return 0;
}
